### PR TITLE
Video comments

### PR DESCRIPTION
**Pull request by dobertRowneySr**
Specs are from #2492.

### CODE REVIEW

File: runtime-modules/content/src/lib.rs [hunk 1]
```diff
@@ -456,6 +467,84 @@ pub struct Person<MemberId> {
...
+    pub cleanup_pay_off: Balance,
+
+    /// When it was created or last edited
+    pub last_edited: BlockNumber,
+
+    /// Overall replies counter
+    pub replies_count: ReplyId,
+
+    /// video associated to the post (instead of the body hash as in the blog module)
+    pub video: VideoId,
```
Comment (bedeho) [inline runtime-modules/content/src/lib.rs]: I don't think we need both a `Post` and a `Reply` representation, correct me if I am wrong.

All we need is one, say `Post`, where its `parent` field is a sum type over a video, e.g. like `ParentId`, if its a post directly to a video itself, or to a prior post on the same video.

File: runtime-modules/content/src/lib.rs [hunk 2]
```diff
@@ -456,6 +467,84 @@ pub struct Person<MemberId> {
...
+    parent_id: ParentId<ReplyId, PostId>,
+
+    /// Pay off by deleting post
+    cleanup_pay_off: Balance,
+
+    /// Last time reply was edited
+    last_edited: BlockNumber,
+
+    /// actual text (hash) of the reply
+    text: Hash,
```
Comment (bedeho) [inline runtime-modules/content/src/lib.rs]: Drop all hash commitments, we are not using this any longer.

File: runtime-modules/content/src/lib.rs [hunk 3]
```diff
@@ -456,6 +467,84 @@ pub struct Person<MemberId> {
...
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct Post_<MemberId, CuratorGroupId, DAOId, Balance, BlockNumber, ReplyId, VideoId> {
+    /// Author of post.
+    pub author: ChannelOwner<MemberId, CuratorGroupId, DAOId>,
+
+    /// Cleanup pay off
+    pub cleanup_pay_off: Balance,
+
+    /// When it was created or last edited
+    pub last_edited: BlockNumber,
```
Comment (bedeho) [inline runtime-modules/content/src/lib.rs]: Check https://github.com/Joystream/joystream/issues/2545, we no longer use this timeout mechanism.

**Review by bedeho — CHANGES_REQUESTED** [review]
Looks like a great start, here are some changes needed

- allow channel owner to turn comment section off or on, and while its off, no comments or reactions on comments are possible. during video creation, the creator can choose to have comment on or off to begin with.
- introduce reactions to the video itself, as distinct from the posts/replies.
- follow this updated scheme that is being introduced in the forum 
https://github.com/Joystream/joystream/issues/2545
- video deletion should include deletion of all posts/replies. This should work in the same way as thread deletion, using witnesses to size of object tree, and unrolling over multiple blocks. Again, we only allow one simultaneous unrolling. In order to traverse the entire tree of posts/replies for a video, an iterable double map with key space `VideoIdxPostId`may be needed to organise the posts/replies.
- Both the channel owner and the content curators can act as moderators. The channel owner can also add a finite number (which must have some hardcoded upper bound) of arbitrary members to the channel which can moderate both video comments and the channel subreddit.
- Can you calculate what the total state footprint, in bytes, of all comments will be as a functino of the number of videos and the average number of comments per video? We need this as a final sanity check on the cost of this feature on the system. If its too expensive, we will need to possible select an off-chain version of this using a metaprotocol version https://github.com/Joystream/joystream/issues/1990. The forum is likely fine as it is, as it won't grow nearly as much as the video comments possibly will.

**Comment by dobertRowneySr** [comment]
The post data structure now looks like this.

``` rust
pub struct Post_<
    MemberId: Default + Clone + Copy,
    CuratorGroupId: Default + Clone + Copy,
    CuratorId: Default + Clone + Copy,
    Balance,
    PostId,
    VideoId,
> {
    /// Author of post.
    pub author: ContentActor<CuratorGroupId, CuratorId, MemberId>,

    /// Cleanup pay off
    pub bloat_bond: Balance,

    /// Overall replies counter
    pub replies_count: PostId,

    /// video associated to the post (instead of the body hash as in the blog module)
    pub post_type: PostType,

    /// parent comment if any
    pub parent_id: Option<PostId>,

    /// video reference
    pub video_reference: VideoId,
}

pub enum PostType {
    VideoPost,
    Comment,
}

```
The `video_reference` field is needed in order to have a quick access to video id the comment is referring to, In order to be inserted into the double map `VideoIdxPostId`. 
The text field works as a description for a `PostType::VideoPost` and a text comment for a `PostType::Comment`. For the moment I allowed only comments from `ContentActor::MemberId(_)`, I don't know If comments by curators should be allowed. Any suggestions @bedeho ?

**Comment by dobertRowneySr** [comment]
## State Footprint computation & estimation
Running `std::mem::size_of` over post in a unit testing scenario (where most of ..Id types are u64 or u32) gives size = 72 bytes.
O(`sizeof(Post_)` x `Averege comments per video` x `Number Of Videos`) = O(`72 x avg_comment_per_video x number_of_video`)

### Estimation using Youtube as a reference
- Comments engagement (= comments/views) for a successful YT channel is at least .5% (source: https://tubularlabs.com/blog/3-metrics-youtube-success/)
- A youtube video (as of now) has on average 8,332 views (source: Google search)
- This means that on average you get 8,332 x .005 = 41.66 < 42 comments per video
- Multiple estimates suggest that there are from 135 to 145 billion videos on youtube (https://www.productmanagementexercises.com/1637/estimate-the-total-number-of-videos-on-youtube)
=> about `72 x 42 x 140,000,000,000 = 4.2336e+14` bytes (as of 2021)

@bedeho does this answer to:
>Can you calculate what the total state footprint, in bytes, of all comments will be as a function of the number of videos and the average number of comments per video?

**Comment by dobertRowneySr** [comment]
As for now I left `cleanup_cost = 1`.
I have the following problem 
- posts are stored in a `double_map`: `VideoId x PostId`
- A `remove_prefix` invocation in a storage double map `N x M` **should** (and maybe I am wrong) have complexity O(log(N) + M) ( [source]( https://www.shawntabrizi.com/substrate/substrate-storage-deep-dive/), [source](https://eth.wiki/fundamentals/patricia-tree))
- cleanup cost for `post` should therefore be sized around `log (next_video_id) + log(post.replies_count)`. This quantity is effectively available at deletion time and not at post creation time, hence I cannot ask for the author to deposit `max( size_of::<Post>() x byte_price, cleanup_cost + margin)` upfront.

So here what I would do:
##### during post creation time
- `bloat_bond = size_of::<Post>() x byte_price` 
- bloat_bond is deposited from post author to post treasury account
##### during cleanup time 
- `bloat_bond = max( size_of::<Post>() x byte_price, cleanup_cost + margin)`
and this quantity (capped at `MaxBloatBond`) it is refunded if actor is the post author, burned otherwise

**Comment by bedeho** [comment]
> A remove_prefix invocation in a storage double map N x M should (and maybe I am wrong) have complexity O(log(N) + M) 

- I could not find where you are getting this from those links? can you be more specific?
- My guess would be that each deletion of a post takes `log(N)` steps, and since you are doing this `M` times, it would be  `Mlog(N)`?
- Your general point here is that database access costs are not constant, but depend on the total size of the current database. This is an interesting point, but surely this should already be an issue confronting pricing normal single database reads, not only these sequences of deletions. Given that the benchmarking approach for Substrate appears to just have a fixed price for reads and writes in all contexts, I wonder if they just choose to pretend that some constant is good enough since `log` grows very slowly eventually. But you should def. dig deeper here, perhaps on the Substrate technical chat.

**Comment by dobertRowneySr** [comment]
>I could not find where you are getting this from those links? can you be more specific?

I was thinking about O(M log N) too, but then I asked on the substrate technical chat and they said it was O(log N + M). I wasn't  fully convinced that is why I expressed caution in the () brackets. Maybe O(M log N) is the cost of M access to a single map and they 
use some optimization for a double map...

**Comment by dobertRowneySr** [comment]
closing in favor of newest Olympia work

**Pull request closed without merging**